### ISSUE TITLE
Unable to determine eth0 interface speed on VPS with if_ plugin

### ISSUE THREAD

**Issue by Gollam** [munin-monitoring/munin#853]
Hi,
The following error happens on my server: 

```
2016/08/02 23:30:02 [INFO]: Starting munin-update
2016/08/02 23:30:02 [INFO] starting work in 31337 for mx.example.com/127.0.0.1:4949.
2016/08/02 23:30:07 [INFO] creating rrd-file for if_eth0->down: '/var/lib/munin/example.com/mx.example.com-if_eth0-down-d.rrd'
2016/08/02 23:30:07 [ERROR] Unable to create '/var/lib/munin/example.com/mx.example.com-if_eth0-down-d.rrd': min must be less than max in DS definition
2016/08/02 23:30:07 [ERROR] In RRD: Error updating /var/lib/munin/example.com/mx.example.com-if_eth0-down-d.rrd: opening '/var/lib/munin/example.com/mx.example.com-if_eth0-down-d.rrd': No such file or directory
2016/08/02 23:30:07 [INFO] creating rrd-file for if_eth0->up: '/var/lib/munin/example.com/mx.example.com-if_eth0-up-d.rrd'
2016/08/02 23:30:07 [ERROR] Unable to create '/var/lib/munin/example.com/mx.example.com-if_eth0-up-d.rrd': min must be less than max in DS definition
2016/08/02 23:30:07 [ERROR] In RRD: Error updating /var/lib/munin/example.com/mx.example.com-if_eth0-up-d.rrd: opening '/var/lib/munin/example.com/mx.example.com-if_eth0-up-d.rrd': No such file or directory
2016/08/02 23:30:07 [INFO]: Munin-update finished for node example.com;mx.example.com (5.01 sec)
```

The problem is that my /sys/class/net/eth0/speed file has the value of **-1**. I see that there was a check for it being positive in the past, and it was reverted: https://github.com/munin-monitoring/munin/commit/290d5ac2be02ced4d09fda68dc561fcf082c9cbf

What should I do about this?

**Comment by sumpfralle** [comment]
Thank you for researching the issue!
Regarding this specific commit: it looks good to me - I think, your issue is caused somewhere else.

Could you please run the following:
```shell
set -x
/etc/munin/plugins/if_eth0
```
This should expose the crucial point, where it all goes wrong :)

**Comment by Gollam** [comment]
I ran the command:

`down.value 48001654771
up.value 4111506105`

It works correctly. The problem is with the _/etc/munin/plugins/if_eth0 config_ command:

`graph_order down up
graph_title eth0 traffic
graph_args --base 1000
graph_vlabel bits in (-) / out (+) per ${graph_period}
graph_category network
graph_info This graph shows the traffic of the eth0 network interface. Please note that the traffic is shown in bits per second, not bytes. IMPORTANT: On 32-bit systems the data source for this plugin uses 32-bit counters, which makes the plugin unreliable and unsuitable for most 100-Mb/s (or faster) interfaces, where traffic is expected to exceed 50 Mb/s over a 5 minute period.  This means that this plugin is unsuitable for most 32-bit production environments. To avoid this problem, use the ip_ plugin instead.  There should be no problems on 64-bit systems running 64-bit kernels.
down.label received
down.type DERIVE
down.graph no
down.cdef down,8,*
down.min 0
up.label bps
up.type DERIVE
up.negative down
up.cdef up,8,*
up.min 0
up.info Traffic of the eth0 interface. Unable to determine interface speed. Please run the plugin as root.`

As you can see, it is unable to determine the speed, and it becomes -1, which is why the RRD file couldn't  be created.

If I run the config as root:

`up.info Traffic of the eth0 interface. Unable to determine interface speed. Please install ethtool, wireless-tools, mii-tool or whatever is appropriate for the interface.`

I installed ethtool, which is also unable to detect the speed:

`+ ethtool eth0
Settings for eth0:
        Supported ports: [ ]
        Supported link modes:   Not reported
        Supported pause frame use: No
        Supports auto-negotiation: No
        Advertised link modes:  Not reported
        Advertised pause frame use: No
        Advertised auto-negotiation: No
        Speed: Unknown!
        Duplex: Unknown! (255)
        Port: Other
        PHYAD: 0
        Transceiver: internal
        Auto-negotiation: off
        Link detected: yes`

This is a KVM based VPS, with virtio network interface.

I also found, that I can define the speed manually for the plugin, with an environment variable:

`  [if_*]
      env.speed 1000
`

But is this the only solution?

**Comment by zorun** [comment]
I also have this issue on a KVM VM running Debian jessie:

```
2017/08/27 14:30:03 [ERROR] Unable to create '/var/lib/munin/domain/host.domain-if_eth0-up-d.rrd': failed to parse data source 600:0:-1000000: min must be less than max in DS definition
```

Here is the output of `munin-run if_eth0 config` on the VM:
```
...
up.min 0
up.max -1000000
up.info Traffic of the eth0 interface. Maximum speed is -1 Mb/s.
down.max -1000000
```

Setting `env.speed` manually works, but the plugin should check for non-sensical values.

**Comment by zorun** [comment]
It is actually fixed since munin 2.0.28 with 120af04a884f9fe1a813ea00397fd386f52c38c7 so this ticket can be closed.

For me upgrading from jessie (2.0.25) to jessie-backports (2.0.33) indeed fixed the issue.

On 2.0.25:
```
up.min 0
up.max -1000000
up.info Traffic of the eth0 interface. Maximum speed is -1 Mb/s.
down.max -1000000
```

This is incorrect.

On 2.0.33:
```
up.min 0
up.info Traffic of the eth0 interface. Unable to determine interface speed. Please install ethtool, wireless-tools, mii-tool or whatever is appropriate for the interface.
```
This is correct, because none of the method (sysfs, ethtool, mii-tool) is able to get the interface speed.

When specifying a max speed manually

```
up.min 0
up.max 10000000000
up.info Traffic of the eth0 interface. Maximum speed is 10000 Mb/s.
down.max 10000000000
```
which is also correct.

So, this ticket can be closed!

**Comment by sumpfralle** [comment]
Good to know - thank you for reporting back!
